openstack/baremetalinspector/v1/introspection/results.go

```diff
@@ -0,0 +1,226 @@
+	}
+	return gophercloud.ExtractNextURL(s.Links)
+}
+
+// UnmarshalJSON trie to convert values for started_at and finished_at from the
+// json response into RFC3339 standard. Since Introspection API can remove the
+// Z from the format, if the conversion fails, it falls back to an RFC3339
+// with no Z format supported by gophercloud.
+func (r *Introspection) UnmarshalJSON(b []byte) error {
+	type tmp Introspection
+	var s struct {
+		tmp
+		StartedAt  string `json:"started_at"`
+		FinishedAt string `json:"finished_at"`
+	}
+
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+
+	*r = Introspection(s.tmp)
+
+	if s.StartedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.StartedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.StartedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.StartedAt = t
+	}
+
+	if s.FinishedAt != "" {
+		t, err := time.Parse(time.RFC3339, s.FinishedAt)
+		if err != nil {
+			t, err = time.Parse(gophercloud.RFC3339NoZ, s.FinishedAt)
+			if err != nil {
+				return err
+			}
+		}
+		r.FinishedAt = t
+	}
+
+	return nil
+}
+
+// GetIntrospectionStatusResult is the response from a Get operation.
+// Call its Extract method to interpret it as an Introspection.
+type GetIntrospectionStatusResult struct {
+	introspectionResult
+}
+
+// StartResult ...
+type StartResult struct {
+	gophercloud.ErrResult
+}
+
+// AbortResult ...
+type AbortResult struct {
+	gophercloud.ErrResult
+}
+
+// Data represents introspection data collected.
+// The format and contents of the stored data depends on the ramdisk used
+// and plugins enabled both in the ramdisk and in inspector itself.
+// This structure has been provided for basic compatibility but it
+// will need extensions
+type Data struct {
+	CPUArch       string                 `json:"cpu_arch"`
+	MACs          []string               `json:"macs"`
+	RootDisk      RootDiskType           `json:"root_disk"`
+	Interfaces    map[string]interface{} `json:"interfaces"`
+	CPUs          int                    `json:"cpus"`
+	BootInterface string                 `json:"boot_interface"`
+	MemoryMB      int                    `json:"memory_mb"`
+	IPMIAddress   string                 `json:"ipmi_address"`
+	Inventory     InventoryType          `json:"inventory"`
+	Error         string                 `json:"error"`
+	LocalGB       int                    `json:"local_gb"`
+	AllInterfaces map[string]interface{} `json:"all_interfaces"`
+}
+
+// RootDiskType ...
+type RootDiskType struct {
+	Rotational         bool        `json:"rotational"`
+	Vendor             string      `json:"vendor"`
+	Name               string      `json:"name"`
+	Hctl               interface{} `json:"hctl"`
+	WwnVendorExtension interface{} `json:"wwn_vendor_extension"`
+	WwnWithExtension   interface{} `json:"wwn_with_extension"`
+	Model              string      `json:"model"`
+	Wwn                interface{} `json:"wwn"`
+	Serial             interface{} `json:"serial"`
```

Comment:
For the above fields, it looks like you're choosing `interface{}` because the test fixture has the value set to `null`? We should try to figure out what kind of core type they are. 

`hctl` looks to be a string: https://github.com/openstack/ironic-python-agent/blob/cfb62c87658e32af373278a8de655367f3b489af/ironic_python_agent/hardware.py#L229-L230

```shell
# ls /sys/block/sdc/device/scsi_device/
2:0:0:1
```

The other fields, which I don't have native access to, also look like strings: https://github.com/openstack/ironic-python-agent/blob/cfb62c87658e32af373278a8de655367f3b489af/ironic_python_agent/tests/unit/test_hardware.py#L898-L906

Reply:
These values can also simply be a null value on the JSON document or possibly even  not exist such as the values noted. See https://github.com/openstack/ironic-python-agent/blob/master/ironic_python_agent/hardware.py#L245

Reply:
It's totally fine if the fields' values can sometimes be `null` / not set. The important part is that the API response is including the field in the response: https://github.com/openstack/ironic-inspector/blob/054f300290c45035bb1cc1b222b2cc8cf26c3002/api-ref/source/samples/api-v1-continue-request.json#L7

On the other hand, if the Ironic Inspector test fixtures are _not_ accurately portraying a _real_ API response (meaning, arbitrary fields are being added just for the sake of a unit test), then that's a bad test fixture and it should not be used in Gophercloud.

In other words: https://github.com/openstack/ironic-inspector/blob/054f300290c45035bb1cc1b222b2cc8cf26c3002/api-ref/source/samples/api-v1-continue-request.json has been used as the basis of the Gophercloud fixture. If this fixture is mocked and not accurate, then it's better to generate fixtures that have actually been pulled from real API calls -- that can make a world of different in how something is implemented in a typed language.

If Ironic Inspector _sometimes_ doesn't set the `wwn_vendor_extension` field in `root_disk`:

```json
  "root_disk": {
    "rotational": true,
    "vendor": "0x1af4",
    "name": "/dev/vda",
    "hctl": null,
    "wwn_with_extension": null,
    "model": "",
    "wwn": null,
    "serial": null,
    "size": 13958643712
  },
```

Then `root_disk` is _absolutely_ a generic `map[string]interface{}`. But because the fixtures are showing `null`, that tells me that Ironic Inspector or the Agent are setting the field somewhere, thus creating a well-formed response.

I hope that makes sense.

Reply:
I had a better look at the code and some definitions are actually deprecated.
Changed code accordingly.